Add class section order check script

ajyoon opened the pull request:
Script to help find errors mentioned in #751.

Actual fixes of the errors exposed by this will be part of a separate PR.

ajyoon commented:
Sample output from the script:
```
Recursively scanning /home/ayoon/Projects/abjad for errors...
[...]
Errors in /home/ayoon/Projects/abjad/experimental/tools/interpolationtools/BreakPointFunction.py:
Lines [309]: BAD HEADER ORDER
===============================================================================
Errors in /home/ayoon/Projects/abjad/experimental/mothballed/ContextMap.py:
Lines [333, 384]: BAD HEADER ORDER
===============================================================================
3187 total files checked.
3048 passed.
139 failed.

```
Also includes a small help blurb activated by `-h`

josiah-wolf-oberholtzer commented:
This looks pretty good, just on a glance-over. I'll try and give it a test-run either tomorrow or within a couple days.

If you're interested, you might try refactoring this into a [CommandlineScript](https://github.com/Abjad/abjad/blob/master/abjad/tools/commandlinetools/CommandlineScript.py) subclass. For example: [StatsScript](https://github.com/Abjad/abjad/blob/master/abjad/tools/commandlinetools/StatsScript.py), [DoctestScript](https://github.com/Abjad/abjad/blob/master/abjad/tools/commandlinetools/DoctestScript.py) and [CleanScript](https://github.com/Abjad/abjad/blob/master/abjad/tools/commandlinetools/CleanScript.py).

That pattern lets Abjad's `ajv` command-line tool locate and execute script classes housed in the `commandlinetools` package automatically.

ajyoon commented:
No problem, I'll work it into a `CommandlineScript` subclass.

josiah-wolf-oberholtzer commented:
Awesome! It should be pretty easy.

Basically:

- Subclass `CommandlineScript`.
- Setup your arguments in a private `def _setup_argument_parser(self, parser):` method.
    - `parser` here is an instance of `ArgumentParser`
- Process the parsed arguments in a private `def _process_args(self, args):` method.
- Define a class-level variable `alias`.
    - The alias names the subcommand for our `ajv` script.
    - e.g. if you choose the alias `check`, you'd be able to call your script via `ajv check`.
- You can define a one-line summary via a class-level variable `short_description`
- You can define a multiline description via a class-level variable `long_description`
    - You can assign your current usage text there, although `argparse` will fill it out with per-argument instructions automatically.

ajyoon commented:
How's this? I also moved the script into `tools/commandlinetools` and refactored it into a `CommandlineScript`. It appears correctly within `ajv` with help and all - otherwise the functionality is all still the same.

josiah-wolf-oberholtzer commented on abjad/tools/commandlinetools/CheckClassSections.py:
```diff
@@ -0,0 +1,169 @@
+    short_description = (
+        'Check the order and contents of class sections in a path.'
+        )
+    long_description = (
```
Drop `Finds` to a new line at the same indentation as `as well as...`. If you wrap the entire string like `long_description = stringtools.normalize(...` here, you'll get cleaner indentation in the `--help` output. Try it for yourself with `ajv check-class-sections --help` to see the inconsistent indentation:

```
17:09 $ ajv check-class-sections --help
usage: check-class-section [-h] [--version] [path]

Check the order and contents of class sections in a path.

positional arguments:
  path        directory tree to be recursed over

optional arguments:
  -h, --help  show this help message and exit
  --version   show program's version number and exit

Finds and lists errors in class section order,
    as well as cases where methods and properties appear
    under incorrect section headers

    If no `path` is given, the current directory is searched.

    Checks that where they appear in classes, the following
    comment headers appear in the given order:

        ### CLASS VARIABLES ###
        ### CONSTRUCTOR ###
        ### INITIALIZER ###
        ### SPECIAL METHODS ###
        ### PRIVATE METHODS ###
        ### PUBLIC METHODS ###
        ### PRIVATE PROPERTIES ###
        ### PUBLIC PROPERTIES ###

    Additionally, this finds cases where methods appear
    under PROPERTIES sections, and vice-versa.
```

josiah-wolf-oberholtzer commented:
One more comment about writing the tests: go ahead and jack most of the logic from the test-case I linked above. I really need to port a lot of that into a dedicated `systemtools.TestCase` class.

You can use the following as a template:

```
# -*- encoding: utf-8 -*-
import argparse
import doctest
import os
import shutil
import unittest
from abjad.tools import commandlinetools
from abjad.tools import stringtools
from abjad.tools import systemtools
try:
    import pathlib
except ImportError:
    import pathlib2 as pathlib
try:
    from StringIO import StringIO
except ImportError:
    from io import StringIO


class TestCase(unittest.TestCase):

    ansi_escape = re.compile(r'\x1b[^m]*m')

    test_path = pathlib.Path(__file__).parent
    subdirectory_path = test_path.joinpath('subdirectory')

    failing_module_path = subdirectory_path.joinpath('BadClass.py')
    failing_module_contents = stringtools.normalize(r'''
    class GoodClass(object):
        <EXAMPLE CODE GOES HERE>
    ''')

    passing_module_path = subdirectory_path.joinpath('GoodClass.py')
    passing_module_contents = stringtools.normalize(r'''
    class BadClass(object):
        <EXAMPLE CODE GOES HERE>
    ''')

    def compare_strings(self, expected, actual):
        example = argparse.Namespace()
        example.want = expected
        output_checker = doctest.OutputChecker()
        flags = (
            doctest.NORMALIZE_WHITESPACE |
            doctest.ELLIPSIS |
            doctest.REPORT_NDIFF
            )
        success = output_checker.check_output(expected, actual, flags)
        if not success:
            diff = output_checker.output_difference(example, actual, flags)
            raise Exception(diff)

    def setUp(self):
        if not self.subdirectory_path.exists():
            self.subdirectory_path.mkdir()
        with open(str(self.failing_module_path), 'w') as file_pointer:
            file_pointer.write(self.failing_module_contents)
        with open(str(self.passing_module_path), 'w') as file_pointer:
            file_pointer.write(self.passing_module_contents)
        self.string_io = StringIO()

    def tearDown(self):
        shutil.rmtree(str(self.subdirectory_path))
        self.string_io.close()

    def test_both(self):
        script = commandlinetools.CheckClassSections()
        command = [str(self.subdirectory_path)]
        with systemtools.TemporaryDirectoryChange(str(self.test_path)):
            with systemtools.RedirectedStreams(stdout=self.string_io):
                with self.assertRaises(SystemExit) as context_manager:
                    script(command)
        assert context_manager.exception.code == 1
        script_output = self.ansi_escape.sub('', self.string_io.getvalue())
        script_output = stringtools.normalize(script_output)
        expected = stringtools.normalize('''
            <EXAMPLE OUTPUT GOES HERE>
        '''.replace('/', os.path.sep))
        self.compare_strings(expected, script_output)
```

ajyoon commented:
How's this? (I see what you mean about the desirability of pulling the boilerplate out of these tests into a higher class...)

josiah-wolf-oberholtzer commented:
Real close. You just need a couple more tests. You currently test that a directory can be recursed over, and that faulty modules in that directory subtree exist and are identified (with correct terminal output and a failing return code of 1). Now test:

- no path specified. `cd` into the subdirectory if necessary. should produce nearly identical output to the current test.
- one failing file: `ajv check-class-sections FailingClass.py` (return code == 1)
- one passing file: `ajv check-class-sections PassingClass.py` (return code == 0)

ajyoon commented:
How's this? I can squash down to 1 commit per file if preferred.

josiah-wolf-oberholtzer commented:
@ajyoon This looks good to me. Obviously, there are million ways to dice things, but you've got good code-coverage, the output seems sane, and the code reads well.

I'm going to merge this into master, and you can start on using the script against the codebase.

If, during usage, you run into situations where you want to extend or fix edge-cases in the script, just submit another PR. I'll try and fast-track those changes in.
